### PR TITLE
Fix PluginName empty when using Ollama/Gemini with AddFromObject (#13516)

### LINKED ISSUE #13516: Bug: The function call plugin name printed is empty
The function call plugin name printed is empty

Image

```

using Microsoft.SemanticKernel;

using Microsoft.SemanticKernel.ChatCompletion;
using Microsoft.SemanticKernel.Connectors.Ollama;
using System.ComponentModel;
using System.Text;

#pragma warning disable SKEXP0110
public class TimePlugin
{
    [KernelFunction, Description("读取指定路径文件的内容，跨平台兼容~目录")]
    public string ReadFile(
        string filePath)
    {
        var ret = File.ReadAllText(filePath, Encoding.UTF8);
        return ret;

    }
    [KernelFunction, Description("This method is used to query the current system time and return it in the format yyyy-MM-dd HH:mm:ss")]
    public string QueryDateTime()
    {
        return DateTime.Now.ToString("yyyy-MM-dd HH:mm:ss");
    }

    [KernelFunction, Description("The purpose of this method is to write the text content to the local computer's target path")]
    public void SaveFile(string filePath, string fileContent)
    {
        System.IO.File.WriteAllText(filePath, fileContent);
    }
}

class Program3
{
    public static async Task Main()
    {
        Console.OutputEncoding = System.Text.Encoding.UTF8;
        var builder = Kernel.CreateBuilder();
        var modelId = "qwen3:4b-instruct-2507-q4_K_M";
        var endpoint = new Uri("http://localhost:11434");

        builder.Services.AddOllamaChatCompletion(modelId, endpoint);

        var kernel = builder.Build();
        //KernelPlugin plugin = kernel.ImportPluginFromType<TimePlugin>();
        var plugin1 = new TimePlugin();
        kernel.Plugins.AddFromObject(plugin1, pluginName:"time");
        var chatCompletionService = kernel.GetRequiredService<IChatCompletionService>();

        string prompt = "读取一下d:\\test123.txt的内容";
        var executionSettings = new OllamaPromptExecutionSettings
        {
            FunctionChoiceBehavior = FunctionChoiceBehavior.Auto(autoInvoke: false)
        };

        var t0 = DateTime.Now;
        var response = await chatCompletionService.GetChatMessageContentAsync(prompt, kernel: kernel, executionSettings: executionSettings);
        var functionCalls = Microsoft.SemanticKernel.FunctionCallContent.GetFunctionCalls(response).ToArray();
        foreach (var functionCall in functionCalls)
        {
            Console.WriteLine($" {functionCall.PluginName}, {functionCall.FunctionName}");//error  have no PluginName
            var functionResult = await functionCall.InvokeAsync(kernel);
            Console.WriteLine($"{functionResult.Result}");
       
        }
        Console.WriteLine($"工具调用数 {functionCalls.Length}");
        Console.Write(response.Content);
    }
}


```

### PR DESCRIPTION
**Pull request by aayushbaluni**
## Description
Fixes #13516

`FunctionCallContentBuilder` used `FunctionName.Parse(fqn)` with the default separator `-` when building `FunctionCallContent` from streaming updates. However, different AI connectors use different separators:
- **Ollama/ChatClient** (FullyQualifiedAIFunction): `_` (e.g. `time_ReadFile`)
- **Gemini**: `_`
- **OpenAI**: `-`
- **FunctionChoiceBehavior**: `.`

When using `kernel.Plugins.AddFromObject(plugin, pluginName: "time")` with Ollama, the model returns `time_ReadFile` but the builder parsed with `-`, producing `pluginName=null` and `functionName="time_ReadFile"`.

## Changes
- Added `ParseFullyQualifiedFunctionName` helper that tries separators `.`, `_`, `-` when parsing
- Added unit tests for underscore and dot separator formats

## Root Cause
`FunctionCallContentBuilder.Build()` called `FunctionName.Parse(fqn)` without a separator, defaulting to `-`. Connectors using `_` or `.` produced unparseable names, leaving `PluginName` empty.

Made with [Cursor](https://cursor.com)

### CODE REVIEW

**Comment by aayushbaluni** [comment]
@microsoft-github-policy-service agree

File: dotnet/src/SemanticKernel.Abstractions/Contents/FunctionCallContentBuilder.cs [hunk 1]
```diff
@@ -170,6 +170,28 @@ public IReadOnlyList<FunctionCallContent> Build()
...
+    /// Parses a fully-qualified function name into plugin and function parts.
+    /// Tries multiple separators (".", "_", "-") since different AI connectors use different formats.
+    /// </summary>
+    private static FunctionName ParseFullyQualifiedFunctionName(string fullyQualifiedName)
+    {
+        foreach (var separator in new[] { ".", "_", "-" })
+        {
+            if (fullyQualifiedName.Contains(separator, StringComparison.Ordinal))
+            {
+                var parsed = FunctionName.Parse(fullyQualifiedName, separator);
```
Comment (chetantoshniwal) [inline dotnet/src/SemanticKernel.Abstractions/Contents/FunctionCallContentBuilder.cs:184]: Bug: This separator-guessing loop will falsely split any bare function name containing `.`, `_`, or `-`. For example, `FunctionName.Parse("get_weather", "_")` produces PluginName=`"get"`, Name=`"weather"`, and the null check on line 185 passes because a non-null PluginName is always produced when the separator exists. Names with multiple underscores (e.g., `my_tools_v2_ReadFile`) split incorrectly as plugin=`my`, function=`tools_v2_ReadFile`. The separator must come from the connector that formatted the name, not be guessed from the string.

```suggestion
    // TODO: accept the separator used by the originating connector instead of guessing.
    private static FunctionName ParseFullyQualifiedFunctionName(string fullyQualifiedName)
    {
        return FunctionName.Parse(fullyQualifiedName);
    }
```

File: dotnet/src/SemanticKernel.Abstractions/Contents/FunctionCallContentBuilder.cs [hunk 2]
```diff
@@ -170,6 +170,28 @@ public IReadOnlyList<FunctionCallContent> Build()
...
+                if (parsed.PluginName is not null)
```
Comment (chetantoshniwal) [inline dotnet/src/SemanticKernel.Abstractions/Contents/FunctionCallContentBuilder.cs:185]: Security: The `parsed.PluginName is not null` check is insufficient—any name containing the separator produces a non-null PluginName. An adversarial model response like `admin_DeleteAllData` would be parsed as plugin=`admin`, function=`DeleteAllData`, potentially routing to an unintended privileged function.

File: dotnet/src/SemanticKernel.Abstractions/Contents/FunctionCallContentBuilder.cs [hunk 3]
```diff
@@ -170,6 +170,28 @@ public IReadOnlyList<FunctionCallContent> Build()
...
+                {
+                    return parsed;
+                }
+            }
+        }
+
```
Comment (chetantoshniwal) [inline dotnet/src/SemanticKernel.Abstractions/Contents/FunctionCallContentBuilder.cs:191]: Including `-` as a candidate separator will silently mis-parse any hyphenated function name as a plugin-qualified call. Hyphens are far more common in identifiers than as SK plugin separators and should not be probed here.

File: dotnet/src/SemanticKernel.Abstractions/Contents/FunctionCallContentBuilder.cs [hunk 4]
```diff
@@ -170,6 +170,28 @@ public IReadOnlyList<FunctionCallContent> Build()
...
+        return FunctionName.Parse(fullyQualifiedName);
+    }
+
```
Comment (chetantoshniwal) [inline dotnet/src/SemanticKernel.Abstractions/Contents/FunctionCallContentBuilder.cs:194]: Structural issue: inserting the new method here splits the XML `<param>` doc comments (lines 170-172) from their target method `TrackStreamingFunctionCallUpdate` (now at line 195), orphaning the parameter documentation. Move this method above the doc block or below `TrackStreamingFunctionCallUpdate`.

File: dotnet/src/SemanticKernel.UnitTests/Contents/FunctionCallBuilder/FunctionCallContentBuilderTests.cs [hunk 5]
```diff
@@ -212,6 +212,48 @@ public void ItShouldCaptureArgumentsDeserializationException(JsonSerializerOptio
         Assert.NotNull(functionCall.Exception);
     }
 
+    [Fact]
+    public void ItShouldParseUnderscoreSeparatorForOllamaAndGeminiConnectors()
```
Comment (chetantoshniwal) [inline dotnet/src/SemanticKernel.UnitTests/Contents/FunctionCallBuilder/FunctionCallContentBuilderTests.cs:216]: The test name `time_ReadFile` is unambiguous because `time` is clearly a plugin. Add a test for a bare function name containing underscores (e.g., `read_file` with no plugin) to verify it is NOT incorrectly split into plugin=`read`, name=`file`.

File: dotnet/src/SemanticKernel.UnitTests/Contents/FunctionCallBuilder/FunctionCallContentBuilderTests.cs [hunk 6]
```diff
@@ -212,6 +212,48 @@ public void ItShouldCaptureArgumentsDeserializationException(JsonSerializerOptio
...
+
+        var functionCalls = sut.Build();
+
+        // Assert
+        var functionCall = Assert.Single(functionCalls);
+        Assert.Equal("time", functionCall.PluginName);
+        Assert.Equal("ReadFile", functionCall.FunctionName);
+    }
+
     private static StreamingChatMessageContent CreateStreamingContentWithFunctionCallUpdate(int choiceIndex, int functionCallIndex, string? callId, string? name, string? arguments, int requestIndex = 0)
```
Comment (chetantoshniwal) [inline dotnet/src/SemanticKernel.UnitTests/Contents/FunctionCallBuilder/FunctionCallContentBuilderTests.cs:257]: Missing test coverage: (1) the hyphen separator path is untested—add a test with e.g. `"time-ReadFile"` asserting PluginName=`"time"`, FunctionName=`"ReadFile"`; (2) the fallback `FunctionName.Parse(fullyQualifiedName)` path is untested—add a test with a plain name like `"ReadFile"` asserting PluginName is null and FunctionName is `"ReadFile"`.
